### PR TITLE
Fix #53: Implement SMS OTP verification

### LINKED ISSUE #53: Implement SMS OTP verification
We should prepare stub methods (sample) for SMS OTP verification method.

### PR DESCRIPTION
**Pull request by romanstrobl**
Note that a new DB table ba_sms_authorization has been added to store SMS authorization codes, messages and maintain SMS authorization status. The bank can use this table or provide their own with some changes to the sample code.

There is one TODO left - template and i18n for the message sent in SMS OTP, right now it is generated by code. We should discuss whether the template should come from Web Auth project or Credential Store / Bank Adapter will have its own resources and how will the message language be switched.

### CODE REVIEW

**Comment by romanstrobl** [comment]
Here are requests for testing:

POST http://localhost:8080/powerauth-credential-server-sample/sms/create
{
	"requestObject": {
		"userId": "12345678",
		"operationName": "payment",
		"operationData": {
			"amount": 100,
			"currency": "CZK",
			"account": "238400856/0300"
		}
	}
}

http://localhost:8080/powerauth-credential-server-sample/sms/verify
{
	"requestObject": {
		"messageId": "76da0913-1201-4f66-9f99-b01f9ad00611",
		"authorizationCode": "61443592"
	}
}

File: powerauth-credential-server-sample/src/main/java/io/getlime/security/powerauth/lib/credentials/controller/SMSAuthorizationController.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+import java.security.SecureRandom;
+import java.util.Date;
+import java.util.UUID;
+
+/**
+ * Controller class which handles SMS OTP authorization.
+ *
+ * @author Roman Strobl, roman.strobl@lime-company.eu
+ */
+@Controller
```
Comment (petrdvorak) [inline powerauth-credential-server-sample/src/main/java/io/getlime/security/powerauth/lib/credentials/controller/SMSAuthorizationController.java]: We are missing base RequestMapping annotation here, and we are not consistent with other controllers:

```java
@Controller
@RequestMapping("/api/auth/sms")
public class ...
```

Next calls then can be called just `/create`, etc.

File: powerauth-credential-server-sample/src/main/java/io/getlime/security/powerauth/lib/credentials/controller/SMSAuthorizationController.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    /**
+     * Create a new SMS OTP authorization message.
+     *
+     * @param request Request data.
+     * @return Response with message ID.
+     */
+    @RequestMapping(value = "/sms/create", method = RequestMethod.POST)
+    public @ResponseBody
```
Comment (petrdvorak) [inline powerauth-credential-server-sample/src/main/java/io/getlime/security/powerauth/lib/credentials/controller/SMSAuthorizationController.java]: Do not break lines between ResponseBody annotation and rest of the method signature, the code looks weird. I guess this is some IDEA formatting issue. You can either write:

```java
@RequestMapping(...)
@ResponseBody
public Object someMethod() {...}
```

... or ...

```java
@RequestMapping(...)
public @ResponseBody Object someMethod() {...}
```

File: powerauth-credential-server-sample/src/main/java/io/getlime/security/powerauth/lib/credentials/controller/SMSAuthorizationController.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            MethodParameter methodParam = new MethodParameter(new Object() {
+            }.getClass().getEnclosingMethod(), 0);
+            throw new MethodArgumentNotValidException(methodParam, result);
+        }
+
+        // messageId is generated as random UUID, it can be overriden to provide a real message identification
+        String messageId = UUID.randomUUID().toString();
+
+        // authorization code generation - to be updated based on application requirements
+        String authorizationCode = String.valueOf((new SecureRandom()).nextInt(90000000) + 10000000);
```
Comment (petrdvorak) [inline powerauth-credential-server-sample/src/main/java/io/getlime/security/powerauth/lib/credentials/controller/SMSAuthorizationController.java]: This is not correct implementation. We are not generating random authorization code. We need to keep the property of "dynamic linking" in place. Basically, we need to make sure that if amount and account number changes, the SMS OTP changes. One way to achieve this is to implement following:

```java
byte[] key = Generator.generateRandomBytes(16);
byte[] signatureBase = amount + "&" + currency + "&" account;
byte[] hash = Mac.hmacSha256(signatureBase, key);
String otp = Util.decimalize(hash);
```

Then, we need to store "key"

You can use PowerAuth crypto, to make your life easier...

Please have a look at:

- https://github.com/lime-company/lime-security-powerauth/blob/ea3c6863b78817e977ed80dc3ce228d3bd8c052e/powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/util/HMACHashUtilities.java#L42
- https://github.com/lime-company/lime-security-powerauth/blob/ea3c6863b78817e977ed80dc3ce228d3bd8c052e/powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/util/SignatureUtils.java#L118
- https://github.com/lime-company/lime-security-powerauth/blob/ea3c6863b78817e977ed80dc3ce228d3bd8c052e/powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/generator/KeyGenerator.java#L108

Comment (petrdvorak) [inline powerauth-credential-server-sample/src/main/java/io/getlime/security/powerauth/lib/credentials/controller/SMSAuthorizationController.java]: (low prio - we can live with this one)

File: powerauth-credential-server-sample/src/main/java/io/getlime/security/powerauth/lib/credentials/controller/SMSAuthorizationController.java (hunk 1 above)
Comment (romanstrobl) [inline powerauth-credential-server-sample/src/main/java/io/getlime/security/powerauth/lib/credentials/controller/SMSAuthorizationController.java]: Yes, I overlooked it. I fixed the RequestMapping in user authentication (login) and user details, please check it.

File: powerauth-credential-server-sample/src/main/java/io/getlime/security/powerauth/lib/credentials/controller/SMSAuthorizationController.java (hunk 2 above)
Comment (romanstrobl) [inline powerauth-credential-server-sample/src/main/java/io/getlime/security/powerauth/lib/credentials/controller/SMSAuthorizationController.java]: Should be fixed. IntelliJ IDEA is the culprit.

File: powerauth-credential-server-sample/src/main/java/io/getlime/security/powerauth/lib/credentials/controller/SMSAuthorizationController.java (hunk 3 above)
Comment (romanstrobl) [inline powerauth-credential-server-sample/src/main/java/io/getlime/security/powerauth/lib/credentials/controller/SMSAuthorizationController.java]: Implemented as requested, only I switched the key and signatureBase parameters in the hmac.hash() method call. The HMACHashUtilities.hash() method requires a key as the first parameter and data as the second parameter, your sample code has reversed parameters. Please check the last commit, we may need to further tweak the authorization code generation. I commented all lines for this code.

**Comment by romanstrobl** [comment]
Updated testing requests:

POST http://localhost:8080/powerauth-credential-server-sample/api/auth/sms/create
```
{
	"requestObject": {
		"userId": "12345678",
		"operationName": "payment",
		"lang": "cs",
		"operationData": {
			"amount": 100,
			"currency": "CZK",
			"account": "238400856/0300"
		}
	}
}
```

POST http://localhost:8080/powerauth-credential-server-sample/api/auth/sms/verify
```
{
	"requestObject": {
		"messageId": "76da0913-1201-4f66-9f99-b01f9ad00611",
		"authorizationCode": "61443592"
	}
}
```

**Pull request merged**